ADD: ui to create multisig

Overtorment opened the pull request.

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/tffycbx50xzw3epderc88115m4

[download apk](http://filestorage.bluewallet.io:1488/1603190024.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/4p23588pc65nxhtv33v6pqr6ur

[download apk](http://filestorage.bluewallet.io:1488/1603208652.apk)

limpbrains commented:
Are you going to change this wallet type selector?
<img width="200" alt="image" src="https://user-images.githubusercontent.com/155891/96711808-d0972f80-13a6-11eb-9e69-81d547e2ccd5.png">

Create button doesn't fit on Android despite the fact that there is enough space (Nexus 5 on appetize0
<img width="200" alt="image" src="https://user-images.githubusercontent.com/155891/96711665-975ebf80-13a6-11eb-8f6d-28d6fb6292b7.png"> <img width="200" alt="image" src="https://user-images.githubusercontent.com/155891/96711890-f290b200-13a6-11eb-98e2-a5d4f47824ed.png">

After key is generated I'm pressing "view" xpub opens very slowly. Why is that ?
On dark theme I can't read "This is cosigner's xpub"

Create is very long. For some time Create button keeps disabled, but at the end it becomes enabled. Better keep it disabled.

Need to be aligned
<img width="200" alt="image" src="https://user-images.githubusercontent.com/155891/96712881-76976980-13a8-11eb-942c-2dd2ce217df8.png">

ncoelho commented:
@limpbrains Yes, the entire UI needs polish and some rework. Let's focus on functionality for now.

limpbrains commented on class/multisig-cosigner.js:
```diff
@@ -0,0 +1,84 @@
+        this._xpub = json.xpub;
+        this._path = json.path;
+        this._cosigners = [true];
+        this._valid = true;
```
I think `return` should be here, after `this._valid = true;`. Why do you need to try coldcard if cobo import was successful?

limpbrains commented on BlueComponents.js:
```diff
@@ -202,6 +202,34 @@ export const BitcoinButton = props => {
+        }}
+      >
+        <View style={{ margin: 16 }}>
+          <Text style={{ color: colors.hdborderColor, fontWeight: 'bold' }}>Multisig Vault</Text>
```
translate please

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/0a7aq17k3ukz7f96mmbrq0q9h8

[download apk](http://filestorage.bluewallet.io:1488/1603297444.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7yb9xeqhz1efj763z85f5jegfr

[download apk](http://filestorage.bluewallet.io:1488/1603311655.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/zcmkgyakkv5m80m6az4hugzerc

[download apk](http://filestorage.bluewallet.io:1488/1603322596.apk)

Overtorment commented:
added `loading` state to second step when its creating a ms. removed `fetchBalance` so its faster.
made `view key` to work with cached value which is filled when key is generated.

limpbrains commented on screen/wallets/addMultisig.js:
```diff
@@ -0,0 +1,325 @@
+
+const styles = StyleSheet.create({
+  modalFlatList: { width: '100%' },
+  whiteBackground: { backgroundColor: BlueCurrentTheme.colors.background },
+  itemNameWrapper: { backgroundColor: BlueCurrentTheme.colors.elevated, flex: 1, flexDirection: 'row', paddingTop: 20, paddingBottom: 0 },
+  textWrapper: { paddingLeft: 10, flex: 1, flexDirection: 'row' },
+  nativeName: { fontSize: 16, color: BlueCurrentTheme.colors.foregroundColor },
+  filteTextWrapper: { color: BlueCurrentTheme.colors.foregroundColor, right: 0, position: 'absolute' },
+  filterText: { fontSize: 16, color: 'gray' },
```
All `BlueCurrentTheme.colors` should come from `useTheme` inside render function

marcosrdz replied:
Fixed.

limpbrains commented on screen/wallets/addMultisig.js:
```diff
@@ -0,0 +1,325 @@
+    margin: 0,
+  },
+  modalContentShort: {
+    backgroundColor: BlueCurrentTheme.colors.elevated,
```
same

marcosrdz replied:
fixed.

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/1fd8qfzk02wwbdzm6t1y8c3wag

[download apk](http://filestorage.bluewallet.io:1488/1603374044.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/khvtbacuna2qa7k2ew583rg1r8

[download apk](http://filestorage.bluewallet.io:1488/1603415061.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/vc42ncnxz0ju93r4dd9t9hhwm0

[download apk](http://filestorage.bluewallet.io:1488/1603453865.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7yuw0j9g5t9ffxkbyeueqf3er4

[download apk](http://filestorage.bluewallet.io:1488/1603479197.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/db7w7vdafk96k23g83wpdndpc8

[download apk](http://filestorage.bluewallet.io:1488/1603578481.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/ntzhu1ph9713gfm4xvb2x9fb5c

[download apk](http://filestorage.bluewallet.io:1488/1603583422.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/wd2fhwd8z2pnp5arnhe9gw9aug

[download apk](http://filestorage.bluewallet.io:1488/1603626103.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7xxefvb37b310jpah78d6x88j4

[download apk](http://filestorage.bluewallet.io:1488/1603674805.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/j62k3tk2ptnk5bqhu99y0d6e50

[download apk](http://filestorage.bluewallet.io:1488/1603676874.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/mhb8fu6h13ecanenym6yvcpekc

[download apk](http://filestorage.bluewallet.io:1488/1603678944.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/8ppqh4dw035jjz1x04qmnpk710

[download apk](http://filestorage.bluewallet.io:1488/1603681939.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/dtccm5d3cp7yvcadjmtf5g6wdg

[download apk](http://filestorage.bluewallet.io:1488/1603683366.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/fx7z1rf735evk2mdckd1eq9tww

[download apk](http://filestorage.bluewallet.io:1488/1603685962.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/da1p5cw0cuzxrgr0mrr7e5uvp4

[download apk](http://filestorage.bluewallet.io:1488/1603718887.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/gthwc8bjmattqdmqrzjkf304cc

[download apk](http://filestorage.bluewallet.io:1488/1603763731.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/wqz57xthbxrmfw5tfta4gzf4hc

[download apk](http://filestorage.bluewallet.io:1488/1603839811.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/vym4dumurz389ev4bc2we638v4

[download apk](http://filestorage.bluewallet.io:1488/1603840598.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/nachmekv18wcdbhdgw86vdc36w

[download apk](http://filestorage.bluewallet.io:1488/1603849503.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/77qpkm8rnrrdbdb99vk41wvph4

[download apk](http://filestorage.bluewallet.io:1488/1603853774.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/6q5xacr2jhk0a98zyj4hj8rc6m

[download apk](http://filestorage.bluewallet.io:1488/1603912525.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/6cf6xemuzf21h5krk7t2ccd1ww

[download apk](http://filestorage.bluewallet.io:1488/1603922575.apk)

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7z75m4qa4kg6tdgcqdanc5pue4

[download apk](http://filestorage.bluewallet.io:1488/1603925071.apk)

marcosrdz commented on screen/wallets/viewEditMultisigCosigners.js:
```diff
@@ -0,0 +1,492 @@
+
+  const m = tempWallet.getM();
+  const n = tempWallet.getN();
+  const format = tempWallet.getFormat();
```
@Overtorment

Overtorment replied:
@marcosrdz refactored it with `useRef`.
long story short, you cant edit global wallet object cause its the object thats used in the app. so to save new wallet you have to create a new object instance, and when user presses  `save` - replace it

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/wqz6u73206u0qtvm6jq12zru5m

[download apk](http://filestorage.bluewallet.io:1488/1603941736.apk)

Overtorment commented on screen/wallets/viewEditMultisigCosigners.js:
```diff
@@ -124,29 +117,36 @@ const ViewEditMultisigCosigners = () => {
-      );
-    }
-    return component;
+    goBack();
```
goBack() wont work here, it navigates back to details of the wallet that doesnt exist anymore, since we deleted it and recreated a new one with a new ID

Overtorment commented on screen/wallets/addMultisig.js:
```diff
@@ -0,0 +1,352 @@
+  },
+});
+
+WalletsAddMultisig.getCurrentFormatReadable = f => {
+  switch (f) {
+    case MultisigHDWallet.FORMAT_P2WSH:
+      return loc.multisig.native_segwit_title;
+    case MultisigHDWallet.FORMAT_P2SH_P2WSH:
+    case MultisigHDWallet.FORMAT_P2SH_P2WSH_ALT:
+      return loc.multisig.wrapped_segwit_title;
+    case MultisigHDWallet.FORMAT_P2SH:
+      return loc.multisig.legacy_title;
+    default:
+      throw new Error('This should never happen');
+  }
+};
```
I dont know how to do that so made it this way

GladosBlueWallet commented:
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7ebrwkpekb17eze3cr4hb45rkm

[download apk](http://filestorage.bluewallet.io:1488/1603974006.apk)

Overtorment commented:
can you guys review again? I intend to merge it, and do a silent release, so its easier to test. 
vault button becomes visible through a backdoor: tap 15 times on btc button

marcosrdz commented on components/DynamicQRCode.js:
```diff
@@ -79,10 +86,15 @@ export class DynamicQRCode extends Component {
     }
   };
 
+  onError = () => {
+    console.log('Data is too large for QR Code.');
+    this.setState({ displayQRCode: false });
+  };
```
This component is not used by itself. Its inside a view with other options such as Sharing the qr data to other apps.

Overtorment replied:
but basically yes, blank screen. this should never happen tho (until it does kek)

marcosrdz replied:
Yeah. But the app would crash without this onError handled. Not the best approach but better than a crash

limpbrains replied:
I mean, you can show some error with `alert` to not to left user uninformed of what is going on

marcosrdz replied:
Yup. That will be addressed.

Overtorment commented on screen/wallets/addMultisigStep2.js:
```diff
@@ -0,0 +1,707 @@
+      }
+
+      for (const existingCosigner of cosigners) {
+        if (existingCosigner[0] === cosigner.getXpub()) return alert('This cosigner is already imported');
```
oops that one was missed
